### ISSUE TITLE
Installation issue: llvm

### ISSUE THREAD

**Issue by behrmann** [LLNL/spack#29342]
### Steps to reproduce the issue

```console
$ spack spec -I llvm
Input spec
--------------------------------
llvm@13%gcc

Concretized
--------------------------------
kslhovh  llvm@13.0.1%gcc@10.2.1+clang~code_signing+compiler-rt~cuda~flang+gold+internal_unwind~ipo~libcxx~link_llvm_dylib+lld+lldb~llvm_dylib~mlir+omp_as_runtime~omp_debug~omp_tsan+polly~python~shared_libs~split_dwarf~z3 build_type=Release targets=none version_suffix=none arch=linux-debian11-skylake
y3wlxwk      ^binutils@2.38%gcc@10.2.1~gas+gold~headers~interwork+ld~libiberty~lto+nls+plugins libs=shared,static arch=linux-debian11-skylake
asesl6c          ^diffutils@3.8%gcc@10.2.1 arch=linux-debian11-skylake
6gzbhqj              ^libiconv@1.16%gcc@10.2.1 libs=shared,static arch=linux-debian11-skylake
wp2cx6c          ^gettext@0.21%gcc@10.2.1+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-debian11-skylake
bztq52k              ^bzip2@1.0.8%gcc@10.2.1~debug~pic+shared arch=linux-debian11-skylake
kqh5uem              ^libxml2@2.9.12%gcc@10.2.1~python arch=linux-debian11-skylake
yckq7g7                  ^pkgconf@1.8.0%gcc@10.2.1 arch=linux-debian11-skylake
wjudasv                  ^xz@5.2.5%gcc@10.2.1~pic libs=shared,static arch=linux-debian11-skylake
7l3ibnl                  ^zlib@1.2.11%gcc@10.2.1+optimize+pic+shared arch=linux-debian11-skylake
befp3xm              ^ncurses@6.2%gcc@10.2.1~symlinks+termlib abi=none arch=linux-debian11-skylake
3xxuheg              ^tar@1.34%gcc@10.2.1 arch=linux-debian11-skylake
w5f4nim      ^cmake@3.22.2%gcc@10.2.1~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-debian11-skylake
jesaas7          ^openssl@1.1.1m%gcc@10.2.1~docs certs=system arch=linux-debian11-skylake
v73srpt              ^perl@5.34.0%gcc@10.2.1+cpanm+shared+threads arch=linux-debian11-skylake
f5tq7pu                  ^berkeley-db@18.1.40%gcc@10.2.1+cxx~docs+stl patches=b231fcc arch=linux-debian11-skylake
7vkbn7d                  ^gdbm@1.19%gcc@10.2.1 arch=linux-debian11-skylake
ikmpaem                      ^readline@8.1%gcc@10.2.1 arch=linux-debian11-skylake
4t7zrhk      ^hwloc@2.7.0%gcc@10.2.1~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=linux-debian11-skylake
hx6potr          ^libpciaccess@0.16%gcc@10.2.1 arch=linux-debian11-skylake
3qly76p              ^libtool@2.4.6%gcc@10.2.1 arch=linux-debian11-skylake
7tmpfit                  ^m4@1.4.19%gcc@10.2.1+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-debian11-skylake
bfk5kkz                      ^libsigsegv@2.13%gcc@10.2.1 arch=linux-debian11-skylake
hn74rjj              ^util-macros@1.19.3%gcc@10.2.1 arch=linux-debian11-skylake
uzl52fj      ^libedit@3.1-20210216%gcc@10.2.1 arch=linux-debian11-skylake
lqirkkf      ^ninja@1.10.2%gcc@10.2.1 arch=linux-debian11-skylake
ov45bff          ^python@3.9.10%gcc@10.2.1+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-debian11-skylake
t2wj6l4              ^expat@2.4.6%gcc@10.2.1+libbsd arch=linux-debian11-skylake
3l3whzy                  ^libbsd@0.11.5%gcc@10.2.1 arch=linux-debian11-skylake
iuvzw6t                      ^libmd@1.0.4%gcc@10.2.1 arch=linux-debian11-skylake
yq6cs6k              ^libffi@3.4.2%gcc@10.2.1 arch=linux-debian11-skylake
4p6awn4              ^sqlite@3.37.2%gcc@10.2.1+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-debian11-skylake
jyy67pa              ^util-linux-uuid@2.37.4%gcc@10.2.1 arch=linux-debian11-skylake
e5tzttp      ^perl-data-dumper@2.173%gcc@10.2.1 arch=linux-debian11-skylake
zsgwvru      ^swig@4.0.2%gcc@10.2.1 arch=linux-debian11-skylake
lgykikt          ^pcre@8.45%gcc@10.2.1~jit+multibyte+utf arch=linux-debian11-skylake
```

### Error message

I can't really post the error message, since it's too long for the issue body. Please have a look at [`spack-build-out.txt`](https://gist.githubusercontent.com/behrmann/6038c4826297b03fdef94ba6165c838f/raw/8ddbe95ab7abecfeef3c0cd767392356fad279ec/spack-build-out.txt)

<details><summary>(shortened) error message</summary><pre>
3 errors found in build log:

  >> 17261    FAILED: bin/lldb-server
  >> 17263    /local_scratch/behrmj87/build/spack/test/spack/opt/spack/linux-debian11-skylake/gcc-10.2.1/binutils-2.38-y3wlxwkc3pjb5mkjoslqh72k6sdhgksl/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/10/../../../x86_64-linux-gnu/libncurses.so.6: undefined reference to `_nc_tiparm@@NCURSES6_TINFO_6.2.current'
  >> 17264    collect2: error: ld returned 1 exit status

See build log for details:
  /tmp/behrmj87/spack-stage/spack-stage-llvm-13.0.1-kslhovhwidx2fmdxyxjeclwap2aiupsh/spack-build-out.txt
</pre></details>

### Information on your system

```console
$ spack debug report
* **Spack:** 0.17.1-1443-6bc2e660e8
* **Python:** 3.9.2
* **Platform:** linux-debian11-skylake
* **Concretizer:** clingo
```

### Additional information

The same thing happens for `llvm@12` (with the same error), but `llvm@11` builds fine. The error occurs with both `-j1` and `-j6` for `spack install`. Disabling lldb (`~lldb`) fixes the build (because the problem is sidestepped). I'm not sure, but this does sound like the error that lead to the report #27085.

The relevant compiler is
```
- compiler:
    spec: gcc@10.2.1
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    flags: {}
    operating_system: debian11
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```

The build information can be found in [this gist](https://gist.github.com/behrmann/6038c4826297b03fdef94ba6165c838f):
- [`spack-build-env.txt`](https://gist.githubusercontent.com/behrmann/6038c4826297b03fdef94ba6165c838f/raw/8ddbe95ab7abecfeef3c0cd767392356fad279ec/spack-build-env.txt)
- [`spack-build-out.txt`](https://gist.githubusercontent.com/behrmann/6038c4826297b03fdef94ba6165c838f/raw/8ddbe95ab7abecfeef3c0cd767392356fad279ec/spack-build-out.txt)

I guess the relevant part of the build output is (linebreaks added for readability)
```console
$ grep /usr/lib/gcc spack-build-out.txte
[3417/5898] cd /tmp/behrmj87/spack-stage/spack-stage-llvm-13.0.1-kslhovhwidx2fmdxyxjeclwap2aiupsh/spack-build-kslhovh/tools/clang/lib/Tooling 
&& /tmp/behrmj87/spack-stage/spack-stage-llvm-13.0.1-kslhovhwidx2fmdxyxjeclwap2aiupsh/spack-build-kslhovh/bin/clang-ast-dump --skip-processing=0 
-I /tmp/behrmj87/spack-stage/spack-stage-llvm-13.0.1-kslhovhwidx2fmdxyxjeclwap2aiupsh/spack-build-kslhovh/lib/clang/13.0.1/include 
-I /tmp/behrmj87/spack-stage/spack-stage-llvm-13.0.1-kslhovhwidx2fmdxyxjeclwap2aiupsh/spack-src/llvm/../clang/include 
-I /tmp/behrmj87/spack-stage/spack-stage-llvm-13.0.1-kslhovhwidx2fmdxyxjeclwap2aiupsh/spack-build-kslhovh/tools/clang/include 
-I /tmp/behrmj87/spack-stage/spack-stage-llvm-13.0.1-kslhovhwidx2fmdxyxjeclwap2aiupsh/spack-build-kslhovh/include 
-I /tmp/behrmj87/spack-stage/spack-stage-llvm-13.0.1-kslhovhwidx2fmdxyxjeclwap2aiupsh/spack-src/llvm/include 
-I /usr/include/c++/10 
-I /usr/include/x86_64-linux-gnu/c++/10 
-I /usr/include/c++/10/backward 
-I /usr/lib/gcc/x86_64-linux-gnu/10/include 
-I /usr/local/include 
-I /usr/include/x86_64-linux-gnu 
-I /usr/include 
--json-output-path /tmp/behrmj87/spack-stage/spack-stage-llvm-13.0.1-kslhovhwidx2fmdxyxjeclwap2aiupsh/spack-build-kslhovh/tools/clang/lib/Tooling/ASTNodeAPI.json

/local_scratch/behrmj87/build/spack/test/spack/opt/spack/linux-debian11-skylake/gcc-10.2.1/binutils-2.38-y3wlxwkc3pjb5mkjoslqh72k6sdhgksl/bin/ld: 
  /usr/lib/gcc/x86_64-linux-gnu/10/../../../x86_64-linux-gnu/libncurses.so.6: undefined reference to `_nc_tiparm@NCURSES6_TINFO_6.2.current'
```
It seems to me like it hits the wrong ncurses (the one from the system), because it tries to grab ncurses relative to gcc. 

cc: @haampie  @trws

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by trws** [comment]
The error looks like it could be from ncurses lacking termlib, was the build using an external ncurses by any chance?

**Comment by haampie** [comment]
You might wanna try the Spack develop branch, there's been some changes to llvm/ncurses since

**Comment by behrmann** [comment]
> The error looks like it could be from ncurses lacking termlib, was the build using an external ncurses by any chance?

I think so, as described above, the build tries to resolve curses relative to the compiler. This was not configured, though.

> You might wanna try the Spack develop branch, there's been some changes to llvm/ncurses since

I'll have a look whether it still persists on develop.

**Comment by omsai** [comment]
Also saw this error linking to ncurses due to the missing symbol `_nc_tiparm@@NCURSES6_TINFO_6.2.current` with with the latest llvm@15 on Debian 11 like @behrmann 

Thanks to @haampie for the suggestion of looking more closely at ncurses. It seems like the bug was fixed on April 20th by Dylan with 7aa1fef5064aa3882ec1d91ab008201e155b77dd. Because my spack-built ncurses that caused the error was built and installed before then, new installations reused that faulty ncurses@6.2 +terminfo.

Building ncurses with the newer package.py and and then using that ncurses to build llvm fixed the linking error.

Feel free to close this issue.
